ships/0005-local-source-upload.md

```diff
@@ -0,0 +1,262 @@
+
+The Shipwright Build Controller works based on [Tekton Workspaces][TektonWorkspaces], and therefore,
+it already defines the /workspace/source mount point for all containers in the build-pod. Tekton also
+adds it's own helper to ["git clone" the repository][ControllerGitClone] in place.
+
+With that said, we can simply stop Shipwright Build-Controller from issuing the "`git clone`", and
+instead, add a [simple helper application][ControllerWaiter] that's able to hold the build-pod
+execution while the data is transferred.
+
+To accomplish this the Build-Controller needs to be able to detect it should replace the "`git
+clone`" procedure by waiting for the upload helper instead. The changes done during research and
+development are [here][ControllerWaiterUsage].
+
+##### BuildRun Changes
+
+The proposed API changes are the following:
+
+```yml
+---
+apiVersion: shipwright.io/v1alpha1
+kind: BuildRun
+metadata: {}
+spec:
+  sources:
+    - name: src
+      type: LocalCopy
+```
+
+The Build Controller will inspect for a type `LocalCopy` on the "`sources"` slice, and therefore
+instead of adding the git clone step, it will instead wait for the user to upload.
+
+The `sources` slice is part of [Remote Artifacts (EP)][RemoteArtifactsEP] and has been worked on, so
+more effort may be needed to create a good API fit.
+
+#### Waiting Mechanism
+
+The container-image built by Shipwright is created by a series of steps running sequentially, those
+steps are executed in the build-pod, running each container and inspecting for return-code and
+errors.
+
+The task at hand, waiting for the users' data transfer, will follow the same basic structure, which
+means a simple application meant to wait for a specific condition, and exit gracefully afterwards.
+Here we may keep checking if a given "lock" file exists, when removed the application exists. It also
+must exit with an error on timeout.
+
+Additionally, we can employ this application to make verifications on the streamed data, like for
+instance checksum, ownership (UID/GID) and more. Another important duty is to have `tar` installed
+and compatible with the activities performed by the CLI.
+
+The implementation written for research and development can be found at [otaviof/waiter][Waiter]
+repository, and it is started by running "`waiter start`". When running it will keep watching the
+file `/var/tmp/waiter.pid`, and when removed, it will exit gracefully. The following example shows it
+running in the container-manager. In the following example the entrypoint receives the argument
+"`start`", and the waiting begins.
+
+```sh
+$ docker run --name=waiter --interactive otaviof/waiter:latest start
+2021/06/01 08:25:24 Saving PID '1' at '/var/tmp/waiter.pid'
+2021/06/01 08:25:24 Attempts '600'...
+2021/06/01 08:25:25 Attempts '599'...
+2021/06/01 08:25:26 Attempts '598'...
+2021/06/01 08:25:27 Done!
+```
+
+When the data transfer is complete, the CLI will execute "`waiter done`", via the same "`kubectl
```

Comment:
have you contemplated the idea on what types of results can be surfaced when the wait is completed. See some examples of what I mean:

- git step -> https://github.com/shipwright-io/build/blob/main/pkg/reconciler/buildrun/resources/sources/git.go#L23-L42
- plain strategy step -> https://github.com/shipwright-io/build/blob/main/samples/buildstrategy/kaniko/buildstrategy_kaniko_cr.yaml#L61

Reply:
Interesting pointers, thank you. I've added wording to describe the  Tekton's `TaskResult` for the `Waiter`, and also termination-log to inform detailed execution results/logs back to Tekton/Kubernetes. Please consider.